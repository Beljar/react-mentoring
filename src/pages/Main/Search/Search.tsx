import { connect } from 'react-redux';
import * as React from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';

import { Button } from 'src/components/ui/Button';
import { Input } from 'src/components/ui/Input/Input';

import { loadMovies, setSearch } from 'src/actions';
import { useParams, useLocation } from 'react-router';
import scss from './styles.scss';

type Props = {
  searchString?: string;
  onSearch?: (searchString) => void;
  onLoad?: () => void;
};

export const Search: React.FC<Props> = ({ searchString = '', onSearch, onLoad }) => {
  const [curSearchString, setCurSearchString] = React.useState('');
  const navigate = useNavigate();
  React.useEffect(() => {
    setCurSearchString(searchString);
  }, [searchString]);
  return (
    <div className={scss.search}>
      <div className={scss.searchBg} />
      <div className={scss.searchImgWrapper}>
        <div className={scss.searchImg} />
      </div>
      <div className={scss.searchContentWrapper}>
        <div className={scss.searchContent}>
          <div className={scss.searchBlock}>
            <h1>FIND YOUR MOVIE</h1>
            <div className={scss.row}>
              <Input
                id="search"
                value={curSearchString}
                placeholder="What do you want to watch?"
                onChange={(value) => {
                  setCurSearchString(value);
                }}
              />
              <Button
                className={scss.searchBtn}
                width={223}
                height={57}
                type="filled"
                onClick={() => {
                  navigate({
                    pathname: '/search',
                    search: `?search=${curSearchString}`,
                  });
                }}
              >
                Search
              </Button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => ({
  searchString: state.request.searchBy === 'title' ? state.request.search : '',
});

const mapDispatchToProps = (dispatch) => ({
  onSearch: (searchString) => dispatch(setSearch(searchString)),
  onLoad: () => dispatch(loadMovies()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Search);
function useQuery() {
  throw new Error('Function not implemented.');
}
