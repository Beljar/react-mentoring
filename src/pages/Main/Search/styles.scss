@import 'src/styles';

.search {
    position: relative;
    width: 100%;
    color: $color-text;
    .search-img {
        width: 100%;
        height: 396px;
        background-image: url('src/img/header-bg.jpg');
        background-position: center;
        mix-blend-mode: normal;
        opacity: 0.2;
        filter: blur(1px);
        -webkit-filter: blur(1px);
        transform: scale(1.1); 
    }  
    .search-img-wrapper {
        position: absolute;
        top: 0;
        overflow: hidden;
        width: 100%;
        height: 396px;
    }
    .search-bg {
        width: 100%;
        height: 396px;
        background-color: $color-black;
    }
    .search-content-wrapper {
        position: absolute;
        top: 66px;
        width: 100%;
        padding: 0px 60px;
        box-sizing: border-box;
    }
    .row {    
        display: flex;
    }
    .search-block {
        width: 960px;
        display: flex;
        flex-direction: column;
        margin: auto;
    }
    h1 {
        font-weight: 300;
        font-size: 40px;
        line-height: 49px;
        margin-top: 35px;
        margin-bottom: 38px;
    }
    .search-btn {
        flex-shrink: 0;
        margin-left: 14px;
    }
}