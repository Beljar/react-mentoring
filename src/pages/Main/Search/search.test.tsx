import * as React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import userEvent from '@testing-library/user-event';
import { Search } from './Search';

let path;

jest.mock('react-router', () => ({
  ...(jest.requireActual('react-router') as object),
  useLocation: () => ({
    pathname: 'localhost:3000/search',
  }),
  useNavigate: () => (val) => {
    path = val;
  },
}));

const PLACEHOLDER = 'What do you want to watch?';

const SEARCH_STRING = 'transformers 7';

const EXPECTED_PATH = { pathname: '/search/transformers 7', search: 'searchBy=title' };

describe('Search component', () => {
  it('renders snapshot', () => {
    const { asFragment } = render(<Search />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders input field', () => {
    const { getByRole } = render(<Search />);
    expect(getByRole('textbox')).toBeInTheDocument();
  });
  it('rendes button', () => {
    const { getByRole } = render(<Search />);
    expect(getByRole('button')).toBeInTheDocument();
  });
  it('renders placeholder', () => {
    const { getByPlaceholderText } = render(<Search />);
    expect(getByPlaceholderText(PLACEHOLDER)).toBeInTheDocument();
  });
  it('displays entered text', async () => {
    const { getByRole, queryByDisplayValue } = render(<Search />);
    const inputEl = getByRole('textbox');
    await userEvent.type(inputEl, SEARCH_STRING);
    expect(queryByDisplayValue(SEARCH_STRING)).toBeInTheDocument();
  });
  it('processes user input', async () => {
    const { getByRole } = render(<Search />);
    const inputEl = getByRole('textbox');
    const buttonEl = getByRole('button');
    await userEvent.type(inputEl, SEARCH_STRING);
    await userEvent.click(buttonEl);
    expect(path).toEqual(EXPECTED_PATH);
  });
});
