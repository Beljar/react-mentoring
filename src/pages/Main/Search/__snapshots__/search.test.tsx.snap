// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <div
      class="searchBg"
    />
    <div
      class="searchImgWrapper"
    >
      <div
        class="searchImg"
      />
    </div>
    <div
      class="searchContentWrapper"
    >
      <div
        class="searchContent"
      >
        <div
          class="searchBlock"
        >
          <h1>
            FIND YOUR MOVIE
          </h1>
          <div
            class="row"
          >
            <input
              class="input"
              id="search"
              placeholder="What do you want to watch?"
              value=""
            />
            <button
              class="searchBtn button filled"
              style="width: 223px; height: 57px;"
              type="button"
            >
              Search
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
